src/features/device-proxy/device-proxy.ts

```diff
@@ -214,22 +289,27 @@ async function requestDevices({
 			}
 			throw new NotFoundError('No online device(s) found');
 		}
-		// Check for device update permission, except for
-		// internal operation of the platform.
+		// Check resource access, unless request is for internal operations of the platform.
 		if (method !== 'GET' && req !== permissions.root) {
-			await Promise.all(
-				deviceIds.map(async (deviceId) => {
-					const res = (await resinApi.request({
-						method: 'POST',
-						url: `device(${deviceId})/canAccess`,
-						body: { action: 'update' },
-					})) as { d?: Array<{ id: number }> };
+			// Find required permission, throw if not found
+			const permission = getPermissionForRequest(method, url);
+			if (permission == null) {
+				throw new errors.ForbiddenError('Unknown endpoint');
+			}
 
-					if (res?.d?.[0]?.id !== deviceId) {
-						throw new errors.ForbiddenError();
-					}
-				}),
-			);
+			// Check permissions depending on the resource type
+			// TODO: Decide if we only want to do application level access checks for shutdown/purge
+			// or if we want to add new auth.ts permissions for application.reboot/restart/etc (and also add to permissionByRequest)
+			// Otherwise we will break things by checking application.update permissions for things like Operators calling application.reboot()
+			if (appId != null && ['shutdown', 'purge'].includes(permission)) {
+				await checkResourceAccess(resinApi, 'application', appId, permission);
+			} else {
+				await Promise.all(
+					deviceIds.map(async (deviceId) =>
+						checkResourceAccess(resinApi, 'device', deviceId, permission),
+					),
+				);
+			}
```

Comment:
I'd prefer to allow application level permission checks as much as possible for performance reasons, but that will require that we add `resin.application.reboot` etc to `auth.ts` and allow those currently able to execute those actions via `resin.device.update` so we don't break anything.